components/gitpod-cli/cmd/validate.go

```diff
@@ -565,58 +567,58 @@ func pipeTask(ctx context.Context, task *api.TaskStatus, supervisor *supervisor.
 }
 
 func listenTerminal(ctx context.Context, task *api.TaskStatus, supervisor *supervisor.SupervisorClient, runLog *logrus.Entry) error {
-	listen, err := supervisor.Terminal.Listen(ctx, &api.ListenTerminalRequest{
-		Alias: task.Terminal,
-	})
+	listen, err := supervisor.Terminal.Listen(ctx, &api.ListenTerminalRequest{Alias: task.Terminal})
 	if err != nil {
 		return err
 	}
 
-	pr, pw := io.Pipe()
-	defer pr.Close()
-	defer pw.Close()
+	var buffer, line bytes.Buffer
 
-	scanner := bufio.NewScanner(pr)
-	const maxTokenSize = 1 * 1024 * 1024 // 1 MB
-	buf := make([]byte, maxTokenSize)
-	scanner.Buffer(buf, maxTokenSize)
+	flushLine := func() {
+		if line.Len() > 0 {
+			runLog.Infof("%s: %s", task.Presentation.Name, line.String())
+			line.Reset()
+		}
+	}
 
-	go func() {
-		defer pw.Close()
-		for {
-			resp, err := listen.Recv()
-			if err != nil {
-				_ = pw.CloseWithError(err)
-				return
+	for {
+		resp, err := listen.Recv()
+		if err != nil {
+			if err == io.EOF {
+				flushLine()
+				return nil
 			}
+			return err
+		}
 
-			title := resp.GetTitle()
-			if title != "" {
-				task.Presentation.Name = title
-			}
+		if title := resp.GetTitle(); title != "" {
+			task.Presentation.Name = title
+		}
 
-			exitCode := resp.GetExitCode()
-			if exitCode != 0 {
-				runLog.Infof("%s: exited with code %d", task.Presentation.Name, exitCode)
-			}
+		if exitCode := resp.GetExitCode(); exitCode != 0 {
+			flushLine()
+			runLog.Infof("%s: exited with code %d", task.Presentation.Name, exitCode)
+		}
 
-			data := resp.GetData()
-			if len(data) > 0 {
-				_, err := pw.Write(data)
-				if err != nil {
-					_ = pw.CloseWithError(err)
-					return
+		if data := resp.GetData(); len(data) > 0 {
+			buffer.Write(data)
+
+			for {
+				r, size := utf8.DecodeRune(buffer.Bytes())
+				if r == utf8.RuneError && size == 0 {
+					break // incomplete character at the end
+				}
+
+				char := buffer.Next(size)
+
+				if r == '\n' || r == '\r' {
```

Comment:
What other (line-)control characters would we need to support as well? ([wikipedia](https://en.wikipedia.org/wiki/Unicode_control_characters)). And: Do we need to handle the windows line ending (`/r/n`) as well? :thinking: 

Otherwise, I think this is a nice approach navigating the difficulties we discussed yesterday, especially in the face of "merging multiple streams". :muscle: